xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -0,0 +1,840 @@
+import io.envoyproxy.envoy.config.listener.v3.Listener;
+import io.envoyproxy.envoy.config.route.v3.RouteConfiguration;
+import io.envoyproxy.envoy.config.route.v3.VirtualHost;
+import io.envoyproxy.envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager;
+import io.envoyproxy.envoy.extensions.filters.network.http_connection_manager.v3.Rds;
+import io.envoyproxy.envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.internal.BackoffPolicy;
+import io.grpc.xds.EnvoyProtoData.DropOverload;
+import io.grpc.xds.EnvoyProtoData.Locality;
+import io.grpc.xds.EnvoyProtoData.LocalityLbEndpoints;
+import io.grpc.xds.EnvoyProtoData.Node;
+import io.grpc.xds.EnvoyProtoData.StructOrError;
+import io.grpc.xds.LoadStatsManager.LoadStatsStore;
+import io.grpc.xds.XdsLogger.XdsLogLevel;
+import java.lang.Thread.UncaughtExceptionHandler;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * XdsClient implementation for client side usages.
+ */
+final class ClientXdsClient extends AbstractXdsClient {
+
+  // Longest time to wait, since the subscription to some resource, for concluding its absence.
+  @VisibleForTesting
+  static final int INITIAL_RESOURCE_FETCH_TIMEOUT_SEC = 15;
+  private static final String TYPE_URL_HTTP_CONNECTION_MANAGER_V2 =
+      "type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2"
+          + ".HttpConnectionManager";
+  private static final String TYPE_URL_HTTP_CONNECTION_MANAGER =
+      "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3"
+          + ".HttpConnectionManager";
+
+  private final Object lock = new Object();
+  private final Map<String, ResourceSubscriber> ldsResourceSubscribers = new HashMap<>();
+  private final Map<String, ResourceSubscriber> rdsResourceSubscribers = new HashMap<>();
+  private final Map<String, ResourceSubscriber> cdsResourceSubscribers = new HashMap<>();
+  private final Map<String, ResourceSubscriber> edsResourceSubscribers = new HashMap<>();
+  private final LoadStatsManager loadStatsManager = new LoadStatsManager();
+  private final LoadReportClient lrsClient;
+  private boolean reportingLoad;
+
+  ClientXdsClient(
+      XdsChannel channel,
+      Node node,
+      ScheduledExecutorService timeService,
+      BackoffPolicy.Provider backoffPolicyProvider,
+      Supplier<Stopwatch> stopwatchSupplier) {
+    super(channel, node, timeService, backoffPolicyProvider, stopwatchSupplier.get());
+    // TODO(chengyuanzhang): delete me after https://github.com/grpc/grpc-java/pull/7528.
```

Comment:
You can delete now, otherwise it will fail to build.

Reply:
Sure. Updated.